crates/spirv-std/src/derivative.rs

```diff
@@ -0,0 +1,98 @@
+    /// Y coordinate. Returns the same result as either [`Self::ddy_fine`] or
+    /// [`Self::ddy_coarse`], selection of which one is dependent on
+    /// external factors.
+    fn ddy(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// Y coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment and its immediate neighbor(s).
+    fn ddy_fine(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// Y coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment’s neighbors, and possibly, but not necessarily,
+    /// includes the value of `Self` for the current fragment. That is, over a
+    /// given area, the implementation can compute Y derivatives in fewer
+    /// unique locations than would be allowed by [`Derivative::ddy_fine`].
+    fn ddy_coarse(self) -> Self;
+    /// Returns the sum of [`Self::ddx`] and [`Self::ddy`] as a
+    /// single operation.
+    fn fwidth(self) -> Self;
+    /// Returns the sum of [`Self::ddx_fine`] and [`Self::ddy_fine`] as a
+    /// single operation.
+    fn fwidth_fine(self) -> Self;
+    /// Returns the sum of [`Self::ddx_coarse`] and [`Self::ddy_coarse`] as a
+    /// single operation.
+    fn fwidth_coarse(self) -> Self;
+}
+
+#[cfg(target_arch = "spirv")]
+macro_rules! deriv_caps {
+    (true) => {
+        asm!("OpCapability DerivativeControl")
+    };
+    (false) => {};
+}
+
+macro_rules! deriv_fn {
+    ($name:ident, $inst:ident, $needs_caps:tt) => {
+        fn $name(self) -> Self {
+            #[cfg(not(target_arch = "spirv"))]
+            panic!(concat!(stringify!($name), " is not supported on the CPU"));
+            #[cfg(target_arch = "spirv")]
+            unsafe {
+                let o;
+                deriv_caps!($needs_caps);
+                asm!(
+                    concat!("{1} = ", stringify!($inst), " typeof{0} {0}"),
+                    in(reg) self,
+                    out(reg) o,
+                );
+                o
+            }
+        }
+    };
+}
+
+macro_rules! deriv_impl {
+    ($ty:ty) => {
+        impl Derivative for $ty {
+            deriv_fn!(ddx, OpDPdx, false);
+            deriv_fn!(ddx_fine, OpDPdxFine, true);
+            deriv_fn!(ddx_coarse, OpDPdxCoarse, true);
+            deriv_fn!(ddy, OpDPdy, false);
+            deriv_fn!(ddy_fine, OpDPdyFine, true);
+            deriv_fn!(ddy_coarse, OpDPdyCoarse, true);
+            deriv_fn!(fwidth, OpFwidth, false);
+            deriv_fn!(fwidth_fine, OpFwidthFine, true);
+            deriv_fn!(fwidth_coarse, OpFwidthCoarse, true);
+        }
+    };
+}
+
+// "must be a scalar or vector of floating-point type. The component width must be 32 bits."
+deriv_impl!(f32);
+// TODO: Fix rustc to support these
+// deriv_impl!(glam::Vec2);
+// deriv_impl!(glam::Vec3);
+// deriv_impl!(glam::Vec4);
```

Comment:
This is offtopic on this PR, but IIRC @eddyb got these working via pointers and `typeof*` in their repo, might be nice to port those back here.